src/argilla/client/feedback/training/frameworks/trl.py

```diff
@@ -218,12 +319,43 @@ def preprocess_function(examples) -> Dict[str, List]:
             )
 
         #  train
-        self._trainer.train()
-        if self._trainer.eval_dataset:
-            self._metrics = self._trainer.evaluate()
-            self._logger.info(self._metrics)
+        if isinstance(self._task, TrainingTaskForPPO):
+            import torch
+            from tqdm import tqdm
+            from trl.core import LengthSampler
+
+            output_length_sampler = LengthSampler(**self.training_args_kwargs["length_sampler_kwargs"])
+            generation_kwargs = self.training_args_kwargs["generation_kwargs"]
+            generation_kwargs["pad_token_id"] = self._transformers_tokenizer.eos_token_id
+            reward_model = self.training_args_kwargs["reward_model"]
+
+            for _, batch in tqdm(enumerate(self._trainer.dataloader)):
+                query_tensors = batch["input_ids"]
+
+                #### Get response from SFT
+                response_tensors = []
+                for query in query_tensors:
+                    gen_len = output_length_sampler()
+                    generation_kwargs["max_new_tokens"] = gen_len
+                    response = self._trainer.generate(query, **generation_kwargs)
+                    response_tensors.append(response.squeeze()[-gen_len:])
+                batch["response"] = [self._transformers_tokenizer.decode(r.squeeze()) for r in response_tensors]
+
+                #### Compute rewards scores
+                texts = [q + r for q, r in zip(batch["query"], batch["response"])]
+                pipe_outputs = reward_model(texts, return_all_scores=True)
+                rewards = [torch.tensor(output[-1]["score"] * len(output)) for output in pipe_outputs]
+
+                #### Run PPO step
+                stats = self._trainer.step(query_tensors, response_tensors, rewards)
+                self._trainer.log_stats(stats, batch, rewards)
         else:
```

Comment:
```suggestion
        else:
            self._metrics = None
```

Reply:
Why should we reintroduce this variable? It's already defined a bit lower right?

Reply:
Yes exactly, sorry I though I had removed this by accident.